### PR TITLE
fix(codebuild): setting Cache.none() renders nothing in the template

### LINKED ISSUE #18165: Codebuild: Removing Cache property does not revert cache to NO_CACHE
### What is the problem?

First setting `Cache = Cache.Local(LocalCacheMode.DOCKER_LAYER),` on a CodeBuild Project and deploying will correctly set the Cache setting on the project.
If you remove the `Cache` set from your source code, the CodeBuild project is not reverted back to `NO_CACHE`. It leaves the previous `DOCKER_LAYER` in place.
This is unexpected as most properties revert to their default option when removed.

### Reproduction Steps

```csharp
            var project = new Project(this, $"{ServiceName}CodeBuild", new ProjectProps
            {
                ProjectName = "Test",
                Cache = Cache.Local(LocalCacheMode.DOCKER_LAYER),
                BuildSpec = BuildSpec.FromObject(new Dictionary<string, object>
                {
                    ["version"] = "0.2",
                    ["phases"] = new Dictionary<string, object>
                    {
                        ["build"] = new Dictionary<string, object>
                        {
                            ["commands"] = new string[] {
                                "rake build[${CODEBUILD_BUILD_NUMBER}]"
                            }
                        }
                    }
                }))
            });

```

### What did you expect to happen?

Removing the `Cache = ` should revert the CodeBuild project to `NO_CACHE`

### What actually happened?

Nothing, removing the property leaves the previous caching property in place.

### CDK CLI Version

2.3.0 (build beaa5b2)

### Framework Version

_No response_

### Node.js Version

v14.16.1

### OS

Windows 11

### Language

.NET

### Language Version

.NET Core 6

### Other information

The reason I spotted this was that I'm in the process of moving to arm64 in our CodeBuild projects. Applying the following change to our CDK stack fails:

image

```
Failed to call UpdateProject, reason: Invalid cache type: local caching is not supported for projects with environment type ARM_CONTAINER and compute type BUILD_GENERAL1_SMALL (Service: AWSCodeBuild; Status Code: 400; Error Code: InvalidInputException; Request ID: 19e3b77c-fcf9-4440-a8ac-1610fb441656; Proxy: null)
        new Project (/tmp/jsii-kernel-moicem/node_modules/@aws-cdk/aws-codebuild/lib/project.js:355:26)
        \_ /tmp/0lk2tlal.3bx/lib/program.js:8432:58
        \_ Kernel._wrapSandboxCode (/tmp/0lk2tlal.3bx/lib/program.js:8860:24)
        \_ Kernel._create (/tmp/0lk2tlal.3bx/lib/program.js:8432:34)
        \_ Kernel.create (/tmp/0lk2tlal.3bx/lib/program.js:8173:29)
        \_ KernelHost.processRequest (/tmp/0lk2tlal.3bx/lib/program.js:9757:36)
        \_ KernelHost.run (/tmp/0lk2tlal.3bx/lib/program.js:9720:22)
        \_ Immediate._onImmediate (/tmp/0lk2tlal.3bx/lib/program.js:9721:46)
        \_ processImmediate (internal/timers.js:464:21)
```

### PR DESCRIPTION
**Pull request by skinny85**
When implementing caching in CodeBuild,
we made the default cache `Cache.none()`,
and, for that reason, do not render anything in the template for that type of cache.
However, that does not work well with the CodeBuild API,
which interprets the lack of a property as the signal to leave it unchanged.
Which means it's not possible currently to disable caching on a Project once it has been enabled once.

Fix this by differentiating between the case of "no Cache has been provided",
and "the none() Cache has been provided".

Closes #18165

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 1]
```diff
@@ -1040,10 +1040,10 @@ export class Project extends ProjectBase {
         : new NoArtifacts());
     const artifactsConfig = artifacts.bind(this, this);
 
-    const cache = props.cache || Cache.none();
+    const cache = props.cache;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts:1043]: The fact that no cache has been provided and the cache mode is left "unchanged" shouldn't even be a mode of operation, so I'm not convinced we should support it.

This PR is fixing the transition:

```ts
  cache: Cache.s3(...)
// ↓
  cache: Cache.none()
```

But the transition:

```ts
  cache: Cache.s3(...)
// ↓
  cache: undefined,
```

is equally valid and should *also* result in caching being turned off.

We should just change the rendering of `Cache.none()`, not change whether or not we render something.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:1043]: Sorry, I'm not sure I understand what the ask here is 😛.

Are you saying we should _always_ render `{ type: 'NO_CACHE' }` in the resulting CloudFormation template, even if the `cache` property of the L2 is `undefined`?

Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:1043]: Discussed offline. Yep, that's what the ask was 🙂. Will implement in the next revision.

**Comment by skinny85** [comment]
@rix0rrr I've changed the default like you asked. As you can see, this resulted in a _lot_ of changes. Let me know if you still want to proceed with this direction - if so, I'll update the PR title and description. Thanks!

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**